File: tools/ghidra_disasm.ts
#!/usr/bin/env node

/**
 * Ghidra Disassembler Tool - TypeScript Wrapper (Refactored for OpenCode)
 * Disassembles binary files using Ghidra's headless analyzer via FastAPI server.
 * 
 * This wrapper communicates with the FastAPI job server (http://tools:8000) instead
 * of spawning child processes. It sends HTTP requests and parses responses.
 */

interface GhidraDisasmConfig {
  binary_path: string;
  function_address?: string;
  function_names?: string[];
  timeout?: number;
}

interface ToolResult {
  status: 'success' | 'error';
  tool_name: string;
  execution_time_ms: number;
  data: any;
  summary: string;
  metadata: {
    ai_next_step?: string;
    [key: string]: any;
  };
}

async function runGhidraDisasm(config: GhidraDisasmConfig): Promise<ToolResult> {
  const startTime = Date.now();

  try {
    // Validate inputs
    if (!config.binary_path) {
      throw new Error('binary_path is required');
    }

    // Build request payload matching FuncRequest model in job_server.py
    const payload: Record<string, any> = {
      binary: config.binary_path,
      timeout: config.timeout || 300
    };

    if (config.function_address) {
      payload.function_address = config.function_address;
    }

    if (config.function_names && config.function_names.length > 0) {
      payload.function_names = config.function_names.join(',');
    }

    // Call FastAPI server at http://tools:8000/disasm
    const apiUrl = 'http://tools:8000/disasm';
    let response;
    try {
      response = await fetch(apiUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout((config.timeout || 300) * 1000)
      });
    } catch (fetchError) {
      const errorMsg = fetchError instanceof Error ? fetchError.message : String(fetchError);
      const executionTime = Date.now() - startTime;
      return {
        status: 'error',
        tool_name: 'ghidra_disasm',
        execution_time_ms: executionTime,
        data: { error: errorMsg },
        summary: `Failed to connect to FastAPI server at ${apiUrl}: ${errorMsg}`,
        metadata: {
          ai_next_step: 'Ensure the tools container is running on http://tools:8000'
        }
      };
    }

    if (!response.ok) {
      let errorDetail = '';
      try {
        const errorBody = await response.json();
        errorDetail = errorBody.detail || JSON.stringify(errorBody);
      } catch {
        errorDetail = await response.text();
      }

      const executionTime = Date.now() - startTime;
      return {
        status: 'error',
        tool_name: 'ghidra_disasm',
        execution_time_ms: executionTime,
        data: { error: `HTTP ${response.status}`, details: errorDetail },
        summary: `FastAPI server returned error: ${response.status} - ${errorDetail}`,
        metadata: {
          ai_next_step: 'Check binary file format and Ghidra installation on tools container'
        }
      };
    }

    let responseData;
    try {
      responseData = await response.json();
    } catch {
      const executionTime = Date.now() - startTime;
      const rawText = await response.text();
      return {
        status: 'error',
        tool_name: 'ghidra_disasm',
        execution_time_ms: executionTime,
        data: { raw: rawText },
        summary: 'Failed to parse FastAPI response as JSON',
        metadata: {
          ai_next_step: 'Check server logs for disassembly errors'
        }
      };
    }

    const executionTime = Date.now() - startTime;

    // Build successful result
    return {
      status: 'success',
      tool_name: 'ghidra_disasm',
      execution_time_ms: executionTime,
      data: responseData,
      summary: `Successfully disassembled ${config.function_address || config.function_names?.join(', ') || 'all functions'} from ${config.binary_path}`,
      metadata: {
        ai_next_step: 'Review assembly instructions; compare with before/after versions for changes'
      }
    };
  } catch (error) {
    const executionTime = Date.now() - startTime;
    const message = error instanceof Error ? error.message : String(error);

    return {
      status: 'error',
      tool_name: 'ghidra_disasm',
      execution_time_ms: executionTime,
      data: { error: message },
      summary: `Ghidra disassembler error: ${message}`,
      metadata: {
        ai_next_step: 'Fix configuration (binary path, function addresses) and retry'
      }
    };
  }
}

// CLI interface for direct invocation
async function main() {
  const args = process.argv.slice(2);
  const config: GhidraDisasmConfig = {
    binary_path: ''
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--binary':
        config.binary_path = args[++i];
        break;
      case '--function-address':
        config.function_address = args[++i];
        break;
      case '--function-names':
        config.function_names = args[++i].split(',');
        break;
      case '--timeout':
        config.timeout = parseInt(args[++i], 10);
        break;
    }
  }

  const result = await runGhidraDisasm(config);
  console.log(JSON.stringify(result, null, 2));
}

if (require.main === module) {
  main().catch((err) => {
    console.error(JSON.stringify({
      status: 'error',
      tool_name: 'ghidra_disasm',
      execution_time_ms: 0,
      data: { error: err.message },
      summary: `Fatal error: ${err.message}`,
      metadata: { ai_next_step: 'Check logs' }
    }, null, 2));
    process.exit(1);
  });
}

export { runGhidraDisasm, GhidraDisasmConfig, ToolResult };

        } catch (e) {
          resolve({
            status: 'error',
            tool_name: 'ghidra_disasm',
            execution_time_ms: executionTime,
            data: { stdout, stderr },
            summary: 'Failed to parse Ghidra disassembly output',
            metadata: {
              ai_next_step: 'Check Python script implementation'
            }
          });
        }
      });
    });
  } catch (error) {
    const executionTime = Date.now() - startTime;
    const message = error instanceof Error ? error.message : String(error);

    return {
      status: 'error',
      tool_name: 'ghidra_disasm',
      execution_time_ms: executionTime,
      data: { error: message },
      summary: `Ghidra disassembler error: ${message}`,
      metadata: {
        ai_next_step: 'Fix configuration and retry'
      }
    };
  }
}

// CLI interface
async function main() {
  const args = process.argv.slice(2);
  const config: GhidraDisasmConfig = {
    binary_path: '',
    output_format: 'json'
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--binary':
        config.binary_path = args[++i];
        break;
      case '--function-address':
        config.function_address = args[++i];
        break;
      case '--function-names':
        config.function_names = args[++i].split(',');
        break;
      case '--output-format':
        config.output_format = args[++i] as 'json' | 'text';
        break;
      case '--ghidra-dir':
        config.ghidra_install_dir = args[++i];
        break;
    }
  }

  const result = await runGhidraDisasm(config);
  console.log(JSON.stringify(result, null, 2));
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: tools/ghidra_disasm.py
#!/usr/bin/env python3

"""
Ghidra Disassembler Tool - Python Orchestrator
Runs Ghidra headless analyzer and Jython script to disassemble binaries
"""

import json
import sys
import os
import argparse
import tempfile
import subprocess
from pathlib import Path
from typing import Dict, List, Any

class GhidraDisassembler:
    def __init__(self, binary_path: str, ghidra_install_dir: str):
        self.binary_path = binary_path
        self.ghidra_install_dir = ghidra_install_dir
        self.jython_script = Path(__file__).parent / 'ghidra_scripts' / 'disassemble.py'

    def run_ghidra_analysis(self, output_file: str, function_filter: str = '') -> Dict[str, Any]:
        """
        Run Ghidra's headless analyzer with Jython script
        """
        try:
            analyze_sh = Path(self.ghidra_install_dir) / 'support' / 'analyzeHeadless'
            
            if not analyze_sh.exists():
                # Try alternative path for Linux
                analyze_sh = Path(self.ghidra_install_dir) / 'support' / 'analyzeHeadless.sh'
            
            if not analyze_sh.exists():
                raise FileNotFoundError(f"analyzeHeadless not found in {self.ghidra_install_dir}")

            # Create temporary project directory
            with tempfile.TemporaryDirectory() as tmpdir:
                project_path = Path(tmpdir) / 'project'
                project_path.mkdir()

                # Build command
                cmd = [
                    str(analyze_sh),
                    str(project_path),           # project directory
                    'ghidra_project',            # project name
                    '-import', self.binary_path,  # import binary
                    '-scriptPath', str(self.jython_script.parent),
                ]

                if function_filter:
                    cmd.extend(['-postScript', 'disassemble.py', function_filter])
                else:
                    cmd.extend(['-postScript', 'disassemble.py'])

                cmd.extend(['-scriptlog', output_file, '-noanalysis'])

                # Run Ghidra
                result = subprocess.run(
                    cmd,
                    capture_output=True,
                    text=True,
                    timeout=300,
                    env={**os.environ, 'GHIDRA_INSTALL_DIR': self.ghidra_install_dir}
                )

                # Parse output from Jython script
                if result.returncode != 0:
                    raise RuntimeError(f"Ghidra analysis failed: {result.stderr}")

                # Read output file if it exists
                if Path(output_file).exists():
                    with open(output_file, 'r') as f:
                        output_content = f.read()
                        # Try to parse as JSON
                        try:
                            return json.loads(output_content)
                        except json.JSONDecodeError:
                            return {'raw_output': output_content}

                return {'message': 'Analysis completed but no output file generated'}

        except Exception as e:
            return {'error': str(e)}

    def analyze(self, function_address: str = '', function_names: List[str] = None) -> Dict[str, Any]:
        """
        Perform disassembly analysis
        """
        try:
            with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as f:
                output_file = f.name

            try:
                # Build function filter
                function_filter = ''
                if function_address:
                    function_filter = f'--address={function_address}'
                elif function_names:
                    function_filter = f'--names={",".join(function_names)}'

                # Run Ghidra
                ghidra_result = self.run_ghidra_analysis(output_file, function_filter)

                return {
                    'disassembly': ghidra_result,
                    'summary': f'Disassembled {self.binary_path}'
                }

            finally:
                if Path(output_file).exists():
                    Path(output_file).unlink()

        except Exception as e:
            return {'error': str(e), 'summary': f'Disassembly failed: {str(e)}'}


def main():
    parser = argparse.ArgumentParser(description='Ghidra Disassembler Tool')
    parser.add_argument('--binary', required=True, help='Path to binary file')
    parser.add_argument('--ghidra-dir', required=True, help='Ghidra installation directory')
    parser.add_argument('--function-address', help='Specific function address to disassemble')
    parser.add_argument('--function-names', help='Comma-separated function names to disassemble')
    parser.add_argument('--output-format', choices=['json', 'text'], default='json', help='Output format')

    args = parser.parse_args()

    try:
        disassembler = GhidraDisassembler(args.binary, args.ghidra_dir)
        
        function_names = None
        if args.function_names:
            function_names = args.function_names.split(',')

        analysis_result = disassembler.analyze(args.function_address, function_names)

        output = {
            'data': analysis_result,
            'summary': analysis_result.get('summary', 'Disassembly completed'),
            'metadata': {
                'binary': args.binary,
                'ghidra_version': 'latest',
                'function_address': args.function_address,
                'function_names': args.function_names,
                'ai_next_step': 'Compare disassembly with before/after versions to identify security fixes'
            }
        }

        if args.output_format == 'json':
            print(json.dumps(output, indent=2))
        else:
            print(f"Ghidra Disassembly Report")
            print(f"========================")
            print(f"Binary: {args.binary}")
            if args.function_address:
                print(f"Function Address: {args.function_address}")
            if args.function_names:
                print(f"Function Names: {args.function_names}")
            print(f"\n{output['summary']}")

    except Exception as e:
        error_output = {
            'error': str(e),
            'summary': f'Ghidra disassembly failed: {str(e)}',
            'metadata': {'ai_next_step': 'Check Ghidra installation and ensure GHIDRA_INSTALL_DIR is set'}
        }
        print(json.dumps(error_output, indent=2))
        sys.exit(1)


if __name__ == '__main__':
    main()
